src/components/ec/cuda/ec_cuda_executor.c

```diff
@@ -0,0 +1,310 @@
+    memcpy(&ee_task->super.args, task_args, sizeof(ucc_ee_executor_task_args_t));
+    switch (task_args->task_type) {
+    case UCC_EE_EXECUTOR_TASK_TYPE_COPY:
+        status = CUDA_FUNC(cudaMemcpyAsync(task_args->bufs[0],
+                                           task_args->bufs[1],
+                                           task_args->count, cudaMemcpyDefault,
+                                           ucc_ec_cuda.stream));
+        if (ucc_unlikely(status != UCC_OK)) {
+            ec_error(&ucc_ec_cuda.super, "failed to start memcpy op");
+            goto free_task;
+        }
+
+        break;
+    case UCC_EE_EXECUTOR_TASK_TYPE_REDUCE:
+        /* temp workaround to avoid code duplication*/
+        status = ucc_mc_reduce(task_args->bufs[1], task_args->bufs[2],
+                               task_args->bufs[0], task_args->count,
+                               task_args->dt, task_args->op,
+                               UCC_MEMORY_TYPE_CUDA);
+        if (ucc_unlikely(status != UCC_OK)) {
+            ec_error(&ucc_ec_cuda.super, "failed to start reduce op");
+            goto free_task;
+        }
+
+        break;
+    default:
+        ec_error(&ucc_ec_cuda.super, "executor operation is not supported");
+        status = UCC_ERR_INVALID_PARAM;
+        goto free_task;
+    }
+
+    status = ucc_ec_cuda_event_post(ucc_ec_cuda.stream, ee_task->event);
+    if (ucc_unlikely(status != UCC_OK)) {
+        goto free_task;
+    }
+
+    *task = &ee_task->super;
+    return UCC_OK;
+
+free_task:
+    ucc_ec_cuda_event_destroy(ee_task->event);
+    ucc_mpool_put(ee_task);
+    return status;
+}
+
+ucc_status_t
+ucc_cuda_executor_interruptible_task_test(const ucc_ee_executor_task_t *task)
+{
+    ucc_ec_cuda_executor_interruptible_task_t *ee_task =
+        ucc_derived_of(task, ucc_ec_cuda_executor_interruptible_task_t);
+    ee_task->super.status = ucc_ec_cuda_event_test(ee_task->event);
+    return ee_task->super.status;
+}
+
+ucc_status_t
+ucc_cuda_executor_interruptible_task_finalize(ucc_ee_executor_task_t *task)
+{
+    ucc_ec_cuda_executor_interruptible_task_t *ee_task =
+        ucc_derived_of(task, ucc_ec_cuda_executor_interruptible_task_t);
+    ucc_status_t status;
+
+    status = ucc_ec_cuda_event_destroy(ee_task->event);
+    ucc_mpool_put(task);
+    return status;
+}
+
+ucc_status_t
+ucc_cuda_executor_persistent_task_post(ucc_ee_executor_t *executor,
+                                       const ucc_ee_executor_task_args_t *task_args,
+                                       ucc_ee_executor_task_t **task)
+{
+    ucc_ec_cuda_executor_t *eee       = ucc_derived_of(executor,
+                                                       ucc_ec_cuda_executor_t);
+    int                     max_tasks = EC_CUDA_CONFIG->exec_max_tasks;
+    ucc_ee_executor_task_t *ee_task;
+
+    if (task_args->task_type == UCC_EE_EXECUTOR_TASK_TYPE_REDUCE) {
+        if (task_args->op != UCC_OP_SUM) {
+            return UCC_ERR_NOT_SUPPORTED;
```

Comment:
question for my understanding: i know we need to implement other ops in executor. Does this mean, that when we land this PR (and until we implement other ops) our triggered allreduce/reduce will only work for sum/float32,64,int32 ?